docs/src/development/proposals/MEP4/README.md

```diff
@@ -1,97 +1,978 @@
+    metal.v2.tenant.delete
+    metal.v2.tenant.create
+    metal.v2.tenant.update
+  Resources:
+    *
+  ```
+- The tenant can now be created for the tenant `new-tenant` even though the logged in tenant is from `metal-stack`.
+  ```
+  $ metalctl tenant create --name new-tenant
+  ...
+  $ metalctl project create --name new-project
+  ...
+  ```
+
+### Custom Role Permissions
+
+A user creates a custom role `ci-builder` and a project token for it:
+
+- The user has permissions to create a role.
+  ```
+  $ metalctl show-permissions
+  Project: 793bb6cd-8b46-479d-9209-0fedca428fe1
+  Roles:
+    metal-project-creator
+    metal-project-owner
+  Permissions:
+    metal.v2.role.list
+    metal.v2.role.get
+    metal.v2.role.create
+    metal.v2.role.update
+    metal.v2.role.delete
+    metal.v2.project.token-list
+    metal.v2.project.token-get
+    metal.v2.project.token-create
+    metal.v2.project.token-revoke
+    ...more typical user permissions...
+  Resources:
+    *
+  ```
+- A user can create a custom role and a role binding for the project. He can't elevate his permissions and can only create roles with permissions the user owns on this project.
+  ```
+  $ metalctl role create --name ci-builder --permissions metal.v2.machine.get,metal.v2.machine.create,metal.v2.machine.delete
+  ```
+- The user can create the project token.
+  ```
+  $ metalctl project token create --role ci-builder
+  <token>
+  ```
+- This command created a role binding:
+  ```
+  $ metalctl rolebinding describe 131e10f0-509f-450f-99d4-3b793cc4db59
+  {
+    "id": "131e10f0-509f-450f-99d4-3b793cc4db59",
+    "name": "project-token",
+    "tenantid": "metal-stack",
+    "projectid": "793bb6cd-8b46-479d-9209-0fedca428fe1",
+    "roles": [
+      {
+        "id": "c3e24647-709e-44da-b551-cbfb1909e328",
+        "name": "ci-builder",
+        "projectid": "793bb6cd-8b46-479d-9209-0fedca428fe1",
+        "permissions": [
+          "metal.v2.machine.get",
+          "metal.v2.machine.create",
+          "metal.v2.machine.delete"
+        ]
+      }
+    ],
+    "userids": [],
+    "projecttokens": ["19c92a80-f5ae-47be-973a-76e47894be8a"],
+    "resources": ["*"],
+    "oidcgroups: []
+  }
+  ```
+- This token can now be used for login, too.
+  ```
+  $ metalctl login --project-token <token>
+  ```
+- The project token has restricted permissions.
+  ```
+  $ metalctl whoami
+  UserId: gerrit
+  Email: gerrit@gerrit.gerrit
+  Tenant: metal-stack
+  Issuer: https://dex.test.io/dex
+  No expiration (project token 19c92a80-f5ae-47be-973a-76e47894be8a)
+
+  $ metalctl show-permissions
+  Project: 793bb6cd-8b46-479d-9209-0fedca428fe1
+  Roles:
+    ci-builder
+  Permissions:
+    metal.v2.machine.get
+    metal.v2.machine.create
+    metal.v2.machine.delete
+  Resources:
+    *
+  ```
+
+❓ Discuss: Should there be a `masterdatactl` (something like an extension of `metalctl`) to maintain tenants, projects, quotas, users, roles and annotations for accounting?
```

Comment:
IMO, it's more convenient to have it in single cmd. Since, if i understood it correctly, permissions for quotas, roles, etc. can be assigned per project to non-admin users.

Reply:
Yes, definitely feels better to have a single interface for the user. 

However, I am unsure where we want to store all these new resources for the described approach... If we put the resources into the masterdata-api, the metal-api does a lot of proxying (boilerplate code, see project service in current metal-api) and maybe it would make more sense to speak with the masterdata-api directly, for performance and maintenance reasons (put a REST gateway in front of it and delegate through ingress?). 

Making the masterdata-api public would require adding OPA to this component as well (or running OPA as a separate service and consume it from multiple components, which is also considerable). Overall API performance should also be considered, as I do not want the user requests to slow down too much just for making auth decisions.